### PR TITLE
Access control

### PR DESCRIPTION
**Pull request by MikeHathaway**
- Removes `recipient` from `addQuoteToken`, `claimCollateral`, `moveQuoteToken`, and `removeQuoteToken`.
- Adds `msg.sender` based access control for modifying Pool liquidity positions
- Rework `PositionManager.sol` into an escrow based contract. All LP tokens are assigned to PM, with the synced `Position` struct tying those LP tokens to a given NFT
- Adds `transferLPTokens` method to `Pool.sol` that allows LP tokens to support optional memorializing of positions. This allows LP token ownership to cleanly be transferred to PM without having to make separate `removeQuoteToken`, and `addQuoteToken` calls for the desired positions to migrate. Since this movement has to occur in the Pool, it required a custom access control method, `approveNewPositionOwner`. I'm currently exploring using `ecrecover` based access control instead, which would offer a smoother UX that didn't require multiple tx on behalf of a memorializing lender.
- Adds OZ.ReentrancyGuard modifier to `decreaseLiquidity<>` methods in PM. LP token amounts to change aren't known until liquidity has been removed, and token transfers have taken place. The Pool methods transferring tokens return the actual equivalent amount of LP tokens, which must be recorded to maintain syncing between the Pool and PositionManager.

### CODE REVIEW

File: src/_test/ERC20Pool/ERC20PoolMove.t.sol [hunk 1]
```diff
@@ -1327,11 +1327,3 @@ contract ERC20PoolMoveQuoteTokenTest is DSTestPlus {
         assertEq(_pool.lpBalance(address(_lender), _p2503), 1_000 * 1e27);
     }
 }
-
-contract ERC20PoolMoveQuoteTokenByRemoveAddTest is ERC20PoolMoveQuoteTokenTest {
```
Comment (grandizzy) [inline src/_test/ERC20Pool/ERC20PoolMove.t.sol:1329]: can we still keep this? it's a proof that move token function produce same results as a remove / add token sequence

File: src/base/Pool.sol [hunk 2]
```diff
@@ -2,6 +2,8 @@
 
 pragma solidity 0.8.14;
 
+import { console } from "@std/console.sol";
```
Comment (grandizzy) [inline src/base/Pool.sol]: is this one still needed?

File: src/base/Pool.sol [hunk 3]
```diff
@@ -68,41 +69,84 @@ abstract contract Pool is IPool, InterestManager, Clone, LenderManager {
...
+        lpBalance[msg.sender][price_] -= lpTokens;
 
         _updateInterestRate(curDebt);
 
         // move quote token amount from pool to lender
-        quoteToken().safeTransfer(recipient_, amount / quoteTokenScale);
-        emit RemoveQuoteToken(recipient_, price_, amount, lup);
+        quoteToken().safeTransfer(msg.sender, amount / quoteTokenScale);
+        emit RemoveQuoteToken(msg.sender, price_, amount, lup);
+        return (amount / quoteTokenScale, lpTokens);
```
Comment (grandizzy) [inline src/base/Pool.sol]: could probably save some gas if we use named variables and perform `amount / quoteTokenScale` operation only once. not a big impact though

File: src/base/Pool.sol [hunk 4]
```diff
@@ -68,41 +69,84 @@ abstract contract Pool is IPool, InterestManager, Clone, LenderManager {
...
+    }
+
+    mapping(address => LpTokenOwnership) lpTokenOwnership;
```
Comment (grandizzy) [inline src/base/Pool.sol]: let's move `lpTokenOwnership` and struct at the top of the file for better visibility / style reasons

File: src/base/Pool.sol [hunk 5]
```diff
@@ -68,41 +69,84 @@ abstract contract Pool is IPool, InterestManager, Clone, LenderManager {
...
+        uint256 tokensTransferred;
+
+        for (uint256 i = 0; i < prices_.length; ) {
+            require(BucketMath.isValidPrice(prices_[i]), "P:TLT:INVALID_PRICE");
+
+            // calculate lp tokens to be moved in the given bucket
+            uint256 tokensToTransfer = lpBalance[owner_][prices_[i]];
+
+            // move lp tokens to the new owners address
+            lpBalance[owner_][prices_[i]] = 0;
```
Comment (grandizzy) [inline src/base/Pool.sol]: does setting to 0 here save more gas than `delete` ? (I have the impression in does but I think it worth double checking)

**Review by grandizzy — COMMENTED** [review]
LGTM, just couple of minor comments and the move quote token test by removing / adding

File: src/base/Pool.sol (hunk 2 above)
Comment (MikeHathaway) [reply src/base/Pool.sol]: Nope, I will remove

File: src/base/Pool.sol (hunk 4 above)
Comment (MikeHathaway) [reply src/base/Pool.sol]: Moved the struct to `IPool.sol` and the mapping to the top of the file.

File: src/base/Pool.sol (hunk 3 above)
Comment (MikeHathaway) [reply src/base/Pool.sol]: Good catch, updated accordingly.

File: src/base/Pool.sol (hunk 5 above)
Comment (MikeHathaway) [reply src/base/Pool.sol]: Good catch, yeah using `delete` saved between several dozen, and several hundred gas depending on the size of the prices array. Updated to use `delete` instead

File: src/_test/ERC20Pool/ERC20PoolMove.t.sol (hunk 1 above)
Comment (MikeHathaway) [reply src/_test/ERC20Pool/ERC20PoolMove.t.sol:1329]: Yes, I had removed it to silence a compiler warning, but was able to resolve by removing override instead. Readded

**Review by EdNoepel — COMMENTED** [review]
LGTM.  Happy to re-review after impending merge conflicts resolved.

**Review by grandizzy — APPROVED** [review]
LGTM!

**Pull request merged**